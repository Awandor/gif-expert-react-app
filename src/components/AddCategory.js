// ========================================
// Functional Component
// ========================================

import { useState } from "react";

import PropTypes from 'prop-types';

const AddCategory = ( { setCategories } ) => {

  const [ inputValue, setInputValue ] = useState( '' ); // Iniciamos con string vacío de lo contrario es undefined y da error

  const handleInputChange = ( event ) => {

    console.log( event.target.value );

    setInputValue( event.target.value ); // sobreescribimos inputValue

  }

  const handleAdd = () => {

    // Para añadir elementos a array hay dos formas

    // setCategories( [...categories, 'Wolverine'] );

    /* setCategories( estadoAnterior =>{

        return [...estadoAnterior, 'Wolverine'];

    } ); */

  };

  const handleSubmit = ( e ) => {

    // No queremos que haga el submit porque eso refresca la página, lo manejaremos de otra forma

    e.preventDefault();

    if ( inputValue.trim().length > 2 ) {

      // Si no tenemos acceso a categories podemos usar callback function que trae categories como argumento

      setCategories( cat => [ inputValue, ...cat ] );

      setInputValue( '' );

    }


  };

  return ( <
    form onSubmit = { handleSubmit } >
    <
    h1 > { inputValue } < /h1> <
    input type = "text"
    value = { inputValue }
    onChange = { handleInputChange }
    /> <
    br / >
    <
    br / >
    <
    button onClick = { handleAdd } > Agregar < /button> <
    /form>
  )

};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;
