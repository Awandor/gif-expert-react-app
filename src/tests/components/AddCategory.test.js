import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';


describe( 'Pruebas en <AddCategory />', () => {

    // const setCategories = () => { };

    // Usamos el método fn() de Jest para decir que es una función y saber varias cosas, si fue llamada, cuántas veces, etc

    const setCategories = jest.fn();

    // Para poner el formulario en blanco antes de cada prueba

    let wrapper;

    beforeEach( ()=>{

        jest.clearAllMocks();

        wrapper = shallow( <AddCategory setCategories={setCategories} /> );
        
    } );


    // Este test se hace con Enzyme

    test( 'debe mostrar <AddCategory /> correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debe de cambiar h1', () => {

        const input = wrapper.find( 'input' );

        const value = 'Hola mundo';

        // input.simulate('change', {}); // {} es el evento que envía handleInputChange

        // input.simulate( 'change', { target:{ value: value } } );
        input.simulate( 'change', { target: { value } } );

        const h1 = wrapper.find( 'h1' ).text().trim();

        console.log( h1 );

        expect( h1 ).toBe( value );

    } );

    test( 'debe de no hacer hacer petición si hay menos de dos caracteres o son espacios vacíos', () => {

        wrapper.find( 'form' ).simulate( 'submit', { preventDefault: ()=>{} } );

        // Ahora setCategories no se debería de haber disparado pues no hemos añadido un término de búsqueda

        console.log( setCategories );

        expect( setCategories ).not.toHaveBeenCalled();

    } );

    test( 'debe de hacer hacer petición y limpiar el buscador', () => {

        const input = wrapper.find( 'input' );

        const value = 'Goku';

        input.simulate( 'change', { target: { value } } );

        wrapper.find( 'form' ).simulate( 'submit', { preventDefault: ()=>{} } );

        // Ahora setCategories se debe disparar

        console.log( setCategories );

        expect( setCategories ).toHaveBeenCalled();

        expect( setCategories ).toHaveBeenCalledTimes( 1 );

        expect( setCategories ).toHaveBeenCalledWith( expect.any( Function ) );

        expect( input.prop( 'value' ) ).toBe( '' );

        

    } );

} );