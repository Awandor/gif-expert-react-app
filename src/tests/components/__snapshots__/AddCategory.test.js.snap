// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> debe mostrar <AddCategory /> correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <br />
  <button
    onClick={[Function]}
  >
    Agregar
  </button>
</form>
`;
